app.py

```diff
@@ -1,104 +1,102 @@
+import streamlit as st 
 from dotenv import load_dotenv
 from PyPDF2 import PdfReader
 from langchain.text_splitter import CharacterTextSplitter
-from langchain.embeddings import OpenAIEmbeddings, HuggingFaceInstructEmbeddings
-from langchain.vectorstores import FAISS
-from langchain.chat_models import ChatOpenAI
+from langchain_community.embeddings.openai import OpenAIEmbeddings
+from langchain_community.embeddings.huggingface import HuggingFaceInstructEmbeddings
+from langchain.vectorstores.faiss import FAISS
+import torch
+from langchain.llms.openai import OpenAI
+from langchain.llms.huggingface_hub import HuggingFaceHub
+from langchain.chat_models.openai import ChatOpenAI
 from langchain.memory import ConversationBufferMemory
-from langchain.chains import ConversationalRetrievalChain
 from htmlTemplates import css, bot_template, user_template
-from langchain.llms import HuggingFaceHub
+from langchain.chains.conversational_retrieval.base import ConversationalRetrievalChain 
 
 def get_pdf_text(pdf_docs):
-    text = ""
+    text = ""  
     for pdf in pdf_docs:
         pdf_reader = PdfReader(pdf)
         for page in pdf_reader.pages:
             text += page.extract_text()
     return text
-
-
+            
 def get_text_chunks(text):
     text_splitter = CharacterTextSplitter(
         separator="\n",
         chunk_size=1000,
         chunk_overlap=200,
-        length_function=len
-    )
+        length_function = len
+        ) 
     chunks = text_splitter.split_text(text)
     return chunks
 
-
 def get_vectorstore(text_chunks):
-    embeddings = OpenAIEmbeddings()
-    # embeddings = HuggingFaceInstructEmbeddings(model_name="hkunlp/instructor-xl")
-    vectorstore = FAISS.from_texts(texts=text_chunks, embedding=embeddings)
-    return vectorstore
+    #embeddings = OpenAIEmbeddings()
+    DEVICE = "cuda:0" if torch.cuda.is_available() else "cpu"
+    embeddings = HuggingFaceInstructEmbeddings(model_name ="hkunlp/instructor-xl",model_kwargs={"device": DEVICE})
+    vector_store = FAISS.from_texts(texts=text_chunks, embedding=embeddings)
+    return vector_store
 
-
-def get_conversation_chain(vectorstore):
+def get_conversation_chain(vector_store):
     llm = ChatOpenAI()
-    # llm = HuggingFaceHub(repo_id="google/flan-t5-xxl", model_kwargs={"temperature":0.5, "max_length":512})
-
-    memory = ConversationBufferMemory(
-        memory_key='chat_history', return_messages=True)
-    conversation_chain = ConversationalRetrievalChain.from_llm(
+    #llm = HuggingFaceHub(repo_id="google/flan-t5-xxl", model_kwargs={"temperature":0.5, "max_length":512})
+    memory = ConversationBufferMemory(memory_key='chat_history', return_messages=True)
+    convesation_chain = ConversationalRetrievalChain.from_llm(
         llm=llm,
-        retriever=vectorstore.as_retriever(),
-        memory=memory
+        retriever=vector_store.as_retreiver(),
+        memory = memory
     )
-    return conversation_chain
+    return convesation_chain
```

Comment:
Should be `return conversation_chain`

Reply:
that's fine he used "covesation_chain" in both places, so should work.
but recommend changing to "conversation_chain"